airflow/providers/google/cloud/operators/cloud_composer.py

```diff
@@ -651,3 +655,146 @@ def execute(self, context: Context):
+
+        cmd, subcommand, parameters = self._parse_cmd_to_args(self.command)
+        execution_cmd_info = hook.execute_airflow_command(
+            project_id=self.project_id,
+            region=self.region,
+            environment_id=self.environment_id,
+            command=cmd,
+            subcommand=subcommand,
+            parameters=parameters,
+            retry=self.retry,
+            timeout=self.timeout,
+            metadata=self.metadata,
+        )
+        execution_cmd_info_dict = ExecuteAirflowCommandResponse.to_dict(execution_cmd_info)
+
+        self.log.info("Command has been started. execution_id=%s", execution_cmd_info_dict["execution_id"])
+
+        if not self.deferrable:
+            result = hook.wait_command_execution_result(
+                project_id=self.project_id,
+                region=self.region,
+                environment_id=self.environment_id,
+                execution_cmd_info=execution_cmd_info_dict,
+                retry=self.retry,
+                timeout=self.timeout,
+                metadata=self.metadata,
+            )
+            result_str = self._merge_cmd_output_result(result)
+            self.log.info("Command execution result: %s", result_str)
+            return result
+        else:
+            self.defer(
+                trigger=CloudComposerAirflowCLICommandTrigger(
+                    project_id=self.project_id,
+                    region=self.region,
+                    environment_id=self.environment_id,
+                    execution_cmd_info=execution_cmd_info_dict,
+                    gcp_conn_id=self.gcp_conn_id,
+                    impersonation_chain=self.impersonation_chain,
+                    poll_interval=self.poll_interval,
+                ),
+                method_name=GOOGLE_DEFAULT_DEFERRABLE_METHOD_NAME,
+            )
+
+    def execute_complete(self, context: Context, event: dict):
+        if event and event["status"] == "error":
+            raise AirflowException(event["message"])
+        result: dict = event["result"]
+        result_str = self._merge_cmd_output_result(result)
+        self.log.info("Command execution result: %s", result_str)
+        return result
+
+    def _parse_cmd_to_args(self, cmd: str) -> tuple:
+        """Parse user command to command, subcommand and parameters."""
+        cmd_dict = shlex.split(cmd)
+        if not cmd_dict:
+            raise AirflowException("The provided command is empty.")
+
+        command = cmd_dict[0] if len(cmd_dict) >= 1 else None
+        subcommand = cmd_dict[1] if len(cmd_dict) >= 2 else None
+        parameters = cmd_dict[2:] if len(cmd_dict) >= 3 else None
+
+        return command, subcommand, parameters
+
+    def _merge_cmd_output_result(self, result):
+        """Merge output to one string."""
+        output = result["output"]
+
+        result_str = "\n"
+        for line_dict in output:
+            result_str += f"{line_dict['content']}\n"
```

Comment:
```suggestion
        result_str = "\n".join(line_dict["content"] for line_dict in output)
```
I'd suggest using a comprehension rather than loop here for performance purposes. It might be sensitive for cases with a very large output.

Here's some experiments for that:
```python
import timeit


messages = [f"message{i}" for i in range(10000)]

def get_message_loop():
    result = "\n"
    for m in messages:
        result += m + "\n"
    return result


def get_message_comprehension():
    return "\n".join(m for m in message)

print(timeit.timeit(get_message_loop, number=100))
print(timeit.timeit(get_message_comprehension, number=100))
```
In my machine with a Python 3.8 the output is:
```
0.15627398784272373
0.06216820399276912
```

Reply:
Really nice, I have added this suggestion to my code.